### PR TITLE
Add SlotFill in Status & Availability panel to match existing hook

### LINKED ISSUE #6299: Add SlotFill in Status & Availability panel to match existing hook.
## Issue Overview
In the legacy Publish meta box, there is an action to allow plugin and theme authors to inject items. This is used by a wide variety of well-known plugins such as Yoast SEO, Edit Flow, and Jetpack’s Publicize module. Currently, there is no counterpart in Gutenberg.

## Possible Solution
Add a slot just above the trash icon in the Status & Visibility panel.

## Screenshots / Video
custom-slot-fill

## Related Issues and/or PRs
https://github.com/WordPress/gutenberg/pull/6300 - also introduces a new SlotFill.

## Todos
- [ ] Tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by ryanwelcher**
## Description
This PR adds a slotFill and is meant as a first pass to fix #6299 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by adamsilverstein** [comment]
This looks good @ryanwelcher , thanks for contributing! I can definately see the use case for adding a slot here and am surprised there isn;t something here already.  It would be good to add some docs for this slot, and possibly tests especially if we already have testing around other slots.

File: edit-post/components/sidebar/plugin-post-status-visibility/index.js
Comment (youknowriad) [inline edit-post/components/sidebar/plugin-post-status-visibility/index.js]: This will include the whole components bundle in the edit-post bundle, to import from other modules, we use `@wordpress/*`. This should be `import { createSlotFille } from '@wordpress/components'`

File: edit-post/components/sidebar/post-status/index.js
Comment (youknowriad) [inline edit-post/components/sidebar/post-status/index.js]: Same as above (external import)

File: edit-post/index.js
Comment (youknowriad) [inline edit-post/index.js]: I'll defer to others on the name and the slot itself cc @gziolo @mtias @aduth

**Comment by adamsilverstein** [comment]
> we don't want to allow plugins to add their own functionalities in Document Settings sidebar.

Where would you suggest plugins add information about the status and visibility of a post - including a published post?

I understand the need to show plugin information during publish flow, however there is also a great use case for displaying document status and visibility information inside the status and visibility area of the editor. In particular, when a plugin has information to add about the status or visibility of the post.

There are many examples, our particular use case is for the Distributor plugin which enables sharing content between WordPress sites. When a post has been shared in this way, that information shows up in the status area:





This information is related to the status of the post you are editing and we would like to show it in Gutenberg in the matching location: 

image

Showing the status information in the publish flow is not helpful the goal of the message is to inform the user of the **distribution status when they open the editor**.  Published posts also need to show this information, and no publish flow is available for these posts. As soon as users open the post edit screen, it should be clear if the post is a distributed (shared) post. Even for drafts, the publish flow _is not available until changes are made to the post, which is disallowed for distributed posts_.

**Comment by jasmussen** [comment]
Thanks so much for your contribution here!

My personal opinion is this is a valid use case, and it is eloquently described by Adam as well.

However I do recall a bunch of discussions around how to build the extensibility, and remember that @mtias had some thoughts about how this is best architected for the future. These discussions, I believe included some thoughts about adding items to existing metaboxes, exactly as is suggested here. 

Alas, I can't recall the exact details, and Matías is AFK this week, but I believe he will be back some time next week. I would prefer that he gets to share his thoughts on this before we merge. But we should be able to have a decision before 2.8.

File: edit-post/components/sidebar/post-status/index.js
Comment (gziolo) [inline edit-post/components/sidebar/post-status/index.js]: I don't think this should be included here. At the moment every item in this Panel is wrapped with `PanelRow`. Instead, we can expose `Fill` which is wrapped with `PanelRow`. In effect, you would add new items as follows:

```js
<PluginPostStatusVisibility>
    My item
< PluginPostStatusVisibility>
```

and `Fill` would work as follows:
```js 
const PluginPostStatusVisibility = ( { children } ) =>  (
    <Fill>
        <PanelRow>
            { children }
        </PanelRow>
    </Fill>
);
```

Alternative solution would be to leave it to plugin developers in case they wanted to add more than one item:

```js
<PluginPostStatusVisibility>
    <PanelRow>My item 1</PanelRow>
    <PanelRow>My item 2</PanelRow>
< PluginPostStatusVisibility>
```

File: edit-post/index.js
Comment (gziolo) [reply edit-post/index.js]: We usually export as `PluginPostStatusVisibility` (Fill) + `PluginPostStatusVisibility.Slot`, but I'm not a big fun of this approach. This one exports them as `PluginPostStatusVisibility.Fill` + `PluginPostStatusVisibility.Slot`. The only benefit of what we have in other places it that Fill is used in many places but Slot usually only in one.

File: edit-post/components/sidebar/post-status/index.js
Comment (ryanwelcher) [reply edit-post/components/sidebar/post-status/index.js]: @gziolo I was thinking the same thing, I would prefer to have a single component. I am new using createSlotFill and am not sure how to implement what you are asking for. Where does 
```
const PluginPostStatusVisibility = ( { children } ) =>  (
    <Fill>
        <PanelRow>
            { children }
        </PanelRow>
    </Fill>
);
```
go?

**Comment by aaronjorbin** [comment]
Along with documentation, I would love to see a simple example added to https://github.com/wordpress/gutenberg-examples 

Maybe something as simple as showing the post ID.

**Comment by gziolo** [comment]
> Along with documentation, I would love to see a simple example added to https://github.com/wordpress/gutenberg-examples
> Maybe something as simple as showing the post ID.

I was thinking we should rather create a plugin that would be included in e2e tests similar to https://github.com/WordPress/gutenberg/blob/master/test/e2e/test-plugins/templates.php but with JS code enqueued. I plan to do something like that for the Sidebar and MenuItem working with Plugins API.

File: edit-post/index.js
Comment (youknowriad) [reply edit-post/index.js]: I think we should only export the Fills

File: edit-post/components/sidebar/post-status/index.js
Comment (gziolo) [reply edit-post/components/sidebar/post-status/index.js]: It can go in here: https://github.com/WordPress/gutenberg/pull/6300/files#diff-ee98bfe279b9987ab8753b1c2fd680f0R7

You will have to juggle a bit with exports.

File: edit-post/components/sidebar/plugin-post-status-visibility/index.js
Comment (aaronjorbin) [inline edit-post/components/sidebar/plugin-post-status-visibility/index.js]: slot, not lot

File: edit-post/components/sidebar/post-status/index.js
Comment (youknowriad) [reply edit-post/components/sidebar/post-status/index.js]: I was only talking about PanelRow, all good now

**Comment by gziolo** [comment]
@ryanwelcher and @adamsilverstein - We need to wait for @mtias to confirm that this change is expected or there is another way that it can be achieved.

In the meantime, do you think it is expected to have other panels extensible in the `Document Settings` sidebar? I wanted to ensure we provide flexible enough solution. If we want to have more panels extensible then we should seek for a more general solution like:
```js
<PluginDocumentSettings panel="post-status-visibility" />
```

**Comment by gziolo** [comment]
@ryanwelcher - sharing the example for the Sidebar for reference, to better picture how it looks in action: https://gist.github.com/gziolo/e2954aa83aa1f823b2b05ca1660c2223. This is also how you would extend `Status & Visibility` if we agree that it is the way to go.

**Comment by adamsilverstein** [comment]
> I wanted to ensure we provide flexible enough solution. If we want to have more panels extensible then we should seek for a more general solution like

I was thinking we need a more general higher level way of creating these slots and wasn't sure how... so in your example `<PluginDocumentSettings panel="status-visibility" />` you would have one of these in each panel, and then plugins would be able to fill them specifically by passing the panel attribute?

**Comment by gziolo** [comment]
@adamsilverstein yes, that's the idea to have matching `Slot` for each of them: `<PluginDocumentSettings.Slot panel="status-visibility" />`. Behind the scenes, it would technically create a few Slot & Fill pairs but would be at least exposed using unified API. It would be much easier to document and use rather than coming up with a name for every possible panel :)

**Comment by adamsilverstein** [comment]
@gziolo in general I like the concept of what you are describing. So looking at the existing panels:



I see the following panels:

> status-visibility
> revisions
> categories
> tags
> featured
> excerpt
> discussion

Do you already have an idea of how this might work behind the scenes?

**Comment by adamsilverstein** [comment]
> as Core developers, we have to be very careful and don't just add slots everywhere because in the end, these will end up abused by extensions even if the said extension doesn't really make sense in this Area

I completely agree with this and wondered about this at first when reading the more general proposal. After considering though, I could imaging a use case for each of the existing panels and I'm sure there are many use cases that I haven't imagined.  

Perhaps we should avoid changing the API for now and simple add slots in spots on a case by case basis where a clear use case can be articulated. We can look to our existing extension points (exposed as `do_action` events in php) for guidance precisely because these actions have been added based on prior use cases and discussion that we should not ignore. We can also look to existing popular plugins for guidance, where we see a valuable use we should consider providing extensibility.

> We should not neglect the fact that adding more and more slots will inevitably create an inconsistent User Experience.

That is a risk, and we need to find a balance between the desire for consistent user experience with the desire to be extensible and useful to the developers who make the platform itself great for users.  Any extension point is an opportunity for unanticipated possibly poor UX. Hopefully we can alleviate that by providing consistent APIs and documenting simple code pattern examples.

**Comment by adamsilverstein** [comment]
@mtias mentioning you here so this one won't slip off your radar.  I'm sure you have many things to review since returning so no hurry ...

**Comment by adamsilverstein** [comment]
@karmatosed we are awaiting design feedback for this ticket (i think), i noticed you removed the needs design feedback a few days ago, was that by mistake or is that the wrong tag to be using here?

**Comment by karmatosed** [comment]
@adamsilverstein I removed as didn't see a design to give feedback on, we usually use that label to mean that. Happy to add back in.

**Comment by mtias** [comment]
Sorry for the delay here!

I think we should move away from **location specific** slots and more towards **purpose specific** ones. Hooks that describe function rather than placement. Location should be a more ephemeral concern — we should be able to iterate on where we place document status, visibility status, etc, without affecting expectations of plugin authors.

The fact "Status and Visibility" are grouped together in a panel is an implementation detail of the current UI. I'm personally not super pleased with its usability, for example, containing a lot of disparate pieces of information and can see it evolving in the future to be something else entirely.

If we add a slot targeting its location alone, it would either prevent us from changing it in the future, or render the slot name confusing to people trying to use it as the visible context has changed. Can we think of a better way to phrase it that more directly hints at the function it will perform?

If we were to move the post status items to the top editor bar as a popover menu (for example, like the table of content "i" button), and the visibility items to the pre-publish panel — how would we expect functionality provided through `PluginPostStatusVisibility` to be treated? We won't know what the intention from someone using it was outside of the location it fulfilled.

When we instead add a slot called `PrePublishStep` we are not tying ourselves with the UI that currently represents it. It could become a modal, a separate page, a sidebar panel, etc, and still make sense from a semantic point of view. Even then, the more granular the better. Something like `PostStatusInfo` or `PostStatusExtraInfo` that solely addresses a specific purpose gives us more freedom to reposition post status information in the future, as the name goes with the function it performs.

When looking at the sidebar as a whole, there are multiple items that we had considered placing elsewhere but didn't yet because of time. The "author" dropdown could be shown more prominently in multi-author sites in the toolbar with a gravatar, "tags" could become a block and be added inline, or bottom of the post, etc, etc. If we make the boxes that contain them the slots we are solidifying the UI. We have the advantage of WP's history to inform the needs to cover, but we should try to not just emulate its previous location but describe the function it articulates as best as we can together with plugin authors.

The main Plugin API with the ellipsis menu, pinned items, and sidebar panel is meant to address the use cases we cannot anticipate while coherently organizing functionality. When instead we need to extend very specific pieces we should be more surgical in identifying the purpose and describing the slots as such.

**Comment by adamsilverstein** [comment]
@mtias thanks for the for the through and nuanced response, this was super helpful. Its great to hear you have thought about this so carefully and your reasoning helps me understand the design goals and overall thinking about slots.

I like your suggestion that we change the name of the slot to `PostStatusInfo` which clearly explains its purpose without tying it to a specific UI position. Can we leave the slot where it is for now, with the understanding that it might change in the future?

**Comment by gziolo** [comment]
> I like your suggestion that we change the name of the slot to `PostStatusInfo` which clearly explains its purpose without tying it to a specific UI position. Can we leave the slot where it is for now, with the understanding that it might change in the future?

I assume it can stay where it is. We should also finally decide if we want to prefix all Slots to be used by plugins with `Plugin`. This PR also needs some documentation with examples how it can be used by plugin developers.

**Comment by jasmussen** [comment]
Might also be good to summarize the gist of Matías' comment in that documentation, so it's transparent from the start that the UI could change.

File: edit-post/components/sidebar/post-status/index.js
Comment (ryanwelcher) [reply edit-post/components/sidebar/post-status/index.js]: @gziolo can you provide some examples of how I would be able to do this - specifically the reference to juggling the exports?

**Comment by ryanwelcher** [comment]
The latest changes address @gziolo feedback to keep the edit-post/index.js file consistent and also only exports the Fill as per @youknowriad. Any feedback is greatly appreciated.

**Comment by adamsilverstein** [comment]
@ryanwelcher looks good! 

The only thing I see as outstanding before merging is updating the documentation, here: https://github.com/WordPress/gutenberg/tree/master/edit-post. As per the  the last two comments above, docs need "examples how it can be used by plugin developers" and "summarize the gist of Matías' comment in that documentation, so it's transparent"

**Comment by gziolo** [comment]
I added https://github.com/WordPress/gutenberg/pull/6300/commits/59601b84df80573d38fddb1cba845e5551e4f76b to simplify `Fill`. Otherwise everything looks great, we just need what @adamsilverstein outlined in the previous comment and we are all set.

**Comment by gziolo** [comment]
Merge with master didn’t go well 🙁

**Comment by ryanwelcher** [comment]
@gziolo it's that kind of day already :) I'm addressing now.

File: edit-post/README.md
Comment (adamsilverstein) [inline edit-post/README.md]: "SlotFills"

**Comment by ryanwelcher** [comment]
@gziolo I've addressed the bad merge with master and added some documentation. Would love some feedback to be sure the docs are clear and complete.

File: edit-post/README.md
Comment (adamsilverstein) [inline edit-post/README.md]: Question: are the items above on this page also slotfills? can you check the codebase and if they are move this header section to the top of the page (eg are PluginSidebar, PluginSidebarMoreMenuItem slots)?

Comment (ryanwelcher) [reply edit-post/README.md]: @adamsilverstein the do have slots defined in them yes. Perhaps we don't need the SlotFills header and I add the note about the implementation vs location under the header for the actual slot.

**Comment by adamsilverstein** [comment]
@gziolo can you please review this again, especially the documentation - I'm really not sure where this belongs, is there one place in the docs that lists all of the slots that are available? 

Thanks!

**Comment by gziolo** [comment]
It’s placed properly. We will have to find a better way to expose it in Gutenberg handbook. At the moment this Readme is only linked from Extensibility docs. I didn’t look closely, I’ll review on Monday. There might be some small stylistic tweaks required but nothing major really. Previous  code examples use JSX and there some other markdown differences. This Slot & Fill idea needs to be better explained in the foreword or sth.

**Review by gziolo — APPROVED** [review]
I addressed my own comments to speed up the process. It is good to go in my opinion. @ryanwelcher and @adamsilverstein hank you for driving this task 🙇

**Pull request merged**
